// elfinfo/dist/src/elf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSymbolsByName = exports.getSymbolByName = exports.getSectionsByName = exports.getSectionByName = exports.fileOffsetToVirtualAddress = exports.fileOffsetToPhysicalAddress = exports.physicalAddressToFileOffset = exports.physicalAddressToVirtual = exports.virtualAddressToFileOffset = exports.virtualAddressToPhysical = exports.getSegmentsAtPhysicalMemoryLocation = exports.getSegmentsAtVirtualMemoryLocation = exports.getSectionsAtPhysicalMemoryLocation = exports.getSectionsAtVirtualMemoryLocation = exports.getSymbolsAtPhysicalMemoryLocation = exports.getSymbolsAtVirtualMemoryLocation = exports.getSegmentForSymbol = exports.getSegmentsForSymbol = exports.getSectionForSymbol = exports.getSectionsForSymbol = exports.getSectionsInSegment = exports.getSymbolsInSegment = exports.getSymbolsInSection = exports.getSymbols = void 0;
const biginthelpers_1 = require("./biginthelpers");
const sections_1 = require("./sections");
function filterSymbolsByVirtualAddress(elf, start, size) {
    const end = biginthelpers_1.add(start, size);
    const symbols = [];
    for (const section of elf.sections) {
        if (sections_1.isSymbolSection(section)) {
            for (const symbol of section.symbols) {
                if (symbol.virtualAddress && symbol.virtualAddress >= start && symbol.virtualAddress < end) {
                    symbols.push(symbol);
                }
            }
        }
    }
    return symbols;
}
/** Get a consolidates array of all the symbols in the file.
 * @param elf the ELF file.
 * @returns an array of symbols.
*/
function getSymbols(elf) {
    const result = [];
    for (const section of elf.sections) {
        if (sections_1.isSymbolSection(section)) {
            for (const sym of section.symbols) {
                result.push(sym);
            }
        }
    }
    return result;
}
exports.getSymbols = getSymbols;
/** Get all the symbols that are addressed inside a given section.
 * @param elf the ELF file.
 * @param {ELFSection | number} sectionOrIndex either the section or the index of the section.
 * @returns {ELFSymbol[]} an array of symbols that are addressed in the section.
 */
function getSymbolsInSection(elf, sectionOrIndex) {
    const section = typeof sectionOrIndex == 'number' ? elf.sections[sectionOrIndex] : sectionOrIndex;
    return filterSymbolsByVirtualAddress(elf, section.addr, section.size);
}
exports.getSymbolsInSection = getSymbolsInSection;
/** Get all the symbols that are addressed inside a given segment.
 * @param {ELFSegment | number} segmentOrIndex either the segment or the index of the segment.
 * @returns {ELFSymbol[]} an array of symbols that are addressed in the segment.
*/
function getSymbolsInSegment(elf, segmentOrIndex) {
    const segment = typeof segmentOrIndex == 'number' ? elf.segments[segmentOrIndex] : segmentOrIndex;
    return filterSymbolsByVirtualAddress(elf, segment.vaddr, segment.memsz);
}
exports.getSymbolsInSegment = getSymbolsInSegment;
/** Get all the section that are addressed inside a given segment.
 * @param {ELFSegment | number} segmentOrIndex either the segment or the index of the segment.
 * @returns {ELFSection[]} an array of sections that are addressed in the segment.
*/
function getSectionsInSegment(elf, segmentOrIndex) {
    const segment = typeof segmentOrIndex == 'number' ? elf.segments[segmentOrIndex] : segmentOrIndex;
    return elf.sections.filter(x => x.addr > segment.vaddr && x.addr < biginthelpers_1.add(segment.vaddr, segment.memsz));
}
exports.getSectionsInSegment = getSectionsInSegment;
/** Get the first section in which a symbol is addressed.
 * @param {Symbol} symbol The symbol
 * @returns {ELFSection[]} an array of sections that contain the symbol.
 */
function getSectionsForSymbol(elf, symbol) {
    const sections = [];
    for (const section of elf.sections) {
        if (symbol.virtualAddress &&
            symbol.virtualAddress >= section.addr &&
            symbol.virtualAddress <= biginthelpers_1.add(section.addr, section.size)) {
            sections.push(section);
        }
    }
    return sections;
}
exports.getSectionsForSymbol = getSectionsForSymbol;
/** Get all sections in which a symbol is addressed.
 * @param {Symbol} symbol The symbol
 * @returns {ELFSection} the first section which contains the symbol.
*/
function getSectionForSymbol(elf, symbol) {
    return getSectionsForSymbol(elf, symbol)[0];
}
exports.getSectionForSymbol = getSectionForSymbol;
/** Get the first segment in which a symbol is addressed.
 * @param {Symbol} symbol The symbol
 * @returns {ELFSection} all segments which contain the symbol.
*/
function getSegmentsForSymbol(elf, symbol) {
    const segments = [];
    for (const segment of elf.segments) {
        if (symbol.virtualAddress &&
            symbol.virtualAddress >= segment.vaddr &&
            symbol.virtualAddress <= biginthelpers_1.add(segment.vaddr, segment.memsz)) {
            segments.push(segment);
        }
    }
    return segments;
}
exports.getSegmentsForSymbol = getSegmentsForSymbol;
/** Get the first segment in which a symbol is addressed.
 * @param {Symbol} symbol The symbol
 * @returns {ELFSection} the first segment which contains the symbol.
*/
function getSegmentForSymbol(elf, symbol) {
    return getSegmentsForSymbol(elf, symbol)[0];
}
exports.getSegmentForSymbol = getSegmentForSymbol;
/** Find all symbols inside that overlap a given virtual memory location.
 * @param {number | bigint} location The virtual memory address.
 * @returns {ELFSymbol[]} an array of symbols that contain the location.
*/
function getSymbolsAtVirtualMemoryLocation(elf, location) {
    const symbols = [];
    for (const section of elf.sections) {
        if (sections_1.isSymbolSection(section)) {
            for (const symbol of section.symbols) {
                if (symbol.size == 0) {
                    if (symbol.virtualAddress === location) {
                        symbols.push(symbol);
                    }
                }
                else {
                    if (symbol.virtualAddress &&
                        location >= symbol.virtualAddress &&
                        location < biginthelpers_1.add(symbol.virtualAddress, symbol.size)) {
                        symbols.push(symbol);
                    }
                }
            }
        }
    }
    return symbols;
}
exports.getSymbolsAtVirtualMemoryLocation = getSymbolsAtVirtualMemoryLocation;
/** Find all symbols inside that overlap a given physical memory location.
 * @param {number | bigint} location The physical memory address.
 * @returns {ELFSymbol[]} an array of symbols that contain the location.
*/
function getSymbolsAtPhysicalMemoryLocation(elf, location) {
    const virtualAddress = physicalAddressToVirtual(elf, location);
    if (virtualAddress) {
        return getSymbolsAtVirtualMemoryLocation(elf, virtualAddress);
    }
    else {
        return [];
    }
}
exports.getSymbolsAtPhysicalMemoryLocation = getSymbolsAtPhysicalMemoryLocation;
/** Get all the sections that overlap a given virtual memory location
 * @param {number | bigint} location The virtual memory address.
 * @returns {ELFSection[]} an array of sections that find the location inside of them.
*/
function getSectionsAtVirtualMemoryLocation(elf, location) {
    const sections = [];
    for (const section of elf.sections) {
        if (location >= section.addr && location < biginthelpers_1.add(section.addr, section.size)) {
            sections.push(section);
        }
    }
    return sections;
}
exports.getSectionsAtVirtualMemoryLocation = getSectionsAtVirtualMemoryLocation;
/** Get all the sections that overlap a given physical memory location
 * @param {number | bigint} location The physical memory address.
 * @returns {ELFSection[]} an array of sections that find the location inside of them.
*/
function getSectionsAtPhysicalMemoryLocation(elf, location) {
    const virtualAddress = physicalAddressToVirtual(elf, location);
    if (virtualAddress) {
        return getSectionsAtVirtualMemoryLocation(elf, virtualAddress);
    }
    else {
        return [];
    }
}
exports.getSectionsAtPhysicalMemoryLocation = getSectionsAtPhysicalMemoryLocation;
/** Get all the segments that overlap a given virtual memory location
 * @param {number | bigint} location The virtual memory address.
 * @returns {ELFSection} all segments which contain the address.
*/
function getSegmentsAtVirtualMemoryLocation(elf, location) {
    const segments = [];
    for (const segment of elf.segments) {
        if (location >= segment.vaddr && location < biginthelpers_1.add(segment.vaddr, segment.memsz)) {
            segments.push(segment);
        }
    }
    return segments;
}
exports.getSegmentsAtVirtualMemoryLocation = getSegmentsAtVirtualMemoryLocation;
/** Get all the segments that overlap a given physical memory location
 * @param {number | bigint} location The physical memory address.
 * @returns {ELFSection} all segments which contain the address.
*/
function getSegmentsAtPhysicalMemoryLocation(elf, location) {
    const segments = [];
    for (const segment of elf.segments) {
        if (location >= segment.paddr && location < biginthelpers_1.add(segment.paddr, segment.filesz)) {
            segments.push(segment);
        }
    }
    return segments;
}
exports.getSegmentsAtPhysicalMemoryLocation = getSegmentsAtPhysicalMemoryLocation;
/** translate a virtual address to a physical address, if possible.
 * @param location The virtual memory address.
 * @returns the physical address.
*/
function virtualAddressToPhysical(elf, location) {
    for (const segment of elf.segments) {
        if (location >= segment.vaddr && location <= biginthelpers_1.add(segment.vaddr, segment.memsz)) {
            const offset = biginthelpers_1.subtract(location, segment.vaddr);
            if (offset < segment.filesz) {
                return biginthelpers_1.add(segment.paddr, offset);
            }
        }
    }
    return undefined;
}
exports.virtualAddressToPhysical = virtualAddressToPhysical;
/** translate a virtual address to an offset in the ELF file, if possible.
 * @param {number | bigint} location The virtual memory address.
 * @returns {number | bigint} the file offset.
*/
function virtualAddressToFileOffset(elf, location) {
    for (const segment of elf.segments) {
        if (location >= segment.vaddr && location < biginthelpers_1.add(segment.vaddr, segment.memsz)) {
            const offset = biginthelpers_1.toNumberSafe(biginthelpers_1.subtract(location, segment.vaddr));
            if (offset < segment.filesz) {
                return segment.offset + offset;
            }
        }
    }
    return undefined;
}
exports.virtualAddressToFileOffset = virtualAddressToFileOffset;
/** translate a physical address to a virtual address.
 * @param {number | bigint} location The physical memory address.
 * @returns {number | bigint} the virtual address.
*/
function physicalAddressToVirtual(elf, location) {
    for (const segment of elf.segments) {
        if (location >= segment.paddr && location < biginthelpers_1.add(segment.paddr, segment.filesz)) {
            const offset = biginthelpers_1.subtract(location, segment.paddr);
            return biginthelpers_1.add(segment.vaddr, offset);
        }
    }
    return undefined;
}
exports.physicalAddressToVirtual = physicalAddressToVirtual;
/** translate a physical address to an offset in the ELF file.
 * @param {number | bigint} location The physical memory address.
 * @returns {number | bigint} the file offset.
*/
function physicalAddressToFileOffset(elf, location) {
    for (const segment of elf.segments) {
        if (location >= segment.paddr && location < biginthelpers_1.add(segment.paddr, segment.filesz)) {
            const offset = biginthelpers_1.toNumberSafe(biginthelpers_1.subtract(location, segment.paddr));
            return segment.offset + offset;
        }
    }
    return undefined;
}
exports.physicalAddressToFileOffset = physicalAddressToFileOffset;
/** translate a file offset to a physical address, if possible.
 * @param {number} location The file offset.
 * @returns {number | bigint} the physical address.
*/
function fileOffsetToPhysicalAddress(elf, location) {
    for (const segment of elf.segments) {
        if (location >= segment.offset && location < biginthelpers_1.add(segment.offset, segment.filesz)) {
            const offset = biginthelpers_1.subtract(location, segment.offset);
            return biginthelpers_1.add(segment.paddr, offset);
        }
    }
    return undefined;
}
exports.fileOffsetToPhysicalAddress = fileOffsetToPhysicalAddress;
/** translate a file offset to a virtual address, if possible.
 * @param {number} location The file offset.
 * @returns {number | bigint} the virtual address.
*/
function fileOffsetToVirtualAddress(elf, location) {
    for (const segment of elf.segments) {
        if (location >= segment.offset && location < biginthelpers_1.add(segment.offset, segment.filesz)) {
            const offset = biginthelpers_1.subtract(location, segment.offset);
            return biginthelpers_1.add(segment.vaddr, offset);
        }
    }
    return undefined;
}
exports.fileOffsetToVirtualAddress = fileOffsetToVirtualAddress;
/** Get the first section that matches the name (case-insensitive).
 * @param {string} sectionName the name of the section to find.
 * @returns {ELFSection} The first section that matches the name
*/
function getSectionByName(elf, sectionName) {
    return getSectionsByName(elf, sectionName)[0];
}
exports.getSectionByName = getSectionByName;
/** Get all sections that matches the name (case-insensitive).
 * @param {string} sectionName the name of the sections to find.
 * @returns {ELFSection[]} an array of sections that match the name.
*/
function getSectionsByName(elf, sectionName) {
    return elf.sections.filter(s => s.name.toUpperCase() == sectionName.toUpperCase());
}
exports.getSectionsByName = getSectionsByName;
/** Get the first symbol that matches the name (case-insensitive).
 * @param {string} symbolName the name of the symbol to find.
 * @returns {ELFSymbol[]} an array of symbols that match the name.
*/
function getSymbolByName(elf, symbolName) {
    for (const section of elf.sections) {
        if (sections_1.isSymbolSection(section)) {
            for (const symbol of section.symbols) {
                if (symbol.name && symbol.name.toUpperCase() == symbolName.toUpperCase()) {
                    return symbol;
                }
            }
        }
    }
    return undefined;
}
exports.getSymbolByName = getSymbolByName;
/** Get all symbols that matches the name (case-insensitive).
 * @param {string} symbolName the name of the symbols to find.
 * @returns {ELFSymbol[]} an array of symbols that match the name.
 *
*/
function getSymbolsByName(elf, symbolName) {
    const matches = [];
    for (const section of elf.sections) {
        if (sections_1.isSymbolSection(section)) {
            for (const symbol of section.symbols) {
                if (symbol.name && symbol.name.toUpperCase() == symbolName.toUpperCase()) {
                    matches.push(symbol);
                }
            }
        }
    }
    return matches;
}
exports.getSymbolsByName = getSymbolsByName;
