// elfinfo/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.open = void 0;
const reader = __importStar(require("./reader"));
const parser_1 = require("./parser");
function isReader(item) {
    return typeof item === 'object' &&
        typeof item.close === 'function' &&
        typeof item.open === 'function' &&
        typeof item.read === 'function' &&
        typeof item.size === 'function' &&
        typeof item.view === 'function';
}
function isFileHandle(item) {
    return typeof item === 'object' &&
        typeof item.fd === 'number' &&
        typeof item.stat === 'function' &&
        typeof item.read === 'function';
}
function isBlob(item) {
    return typeof item === 'object' &&
        typeof item.size === 'number' &&
        typeof item.arrayBuffer === 'function';
}
const defaultOptions = {
    readSymbolData: false
};
/**
 * Parse an ELF file.
 * @summary Parsing will be async if a path, blob, or file handle is specified and synchronous if an array or buffer is specified.
 * @param {any} input the path to the ELF file, or the data for the file.
 * @param {function} [callback] When specified, this will be called after the file is done parsing.
 * @returns {Promise<ELFOpenResult>} a result indicating the success or failure of parsing and the data for the ELF file.
 */
function open(input, options, callback) {
    let promise;
    if (!options) {
        options = defaultOptions;
    }
    if (input instanceof Uint8Array) {
        promise = parser_1.readElf(reader.buffer(input), options);
    }
    else if (input instanceof ArrayBuffer) {
        promise = parser_1.readElf(reader.buffer(input), options);
    }
    else if (input instanceof Array) {
        promise = parser_1.readElf(reader.array(input), options);
    }
    else if (isReader(input)) {
        promise = parser_1.readElf(input, options);
    }
    else if (typeof input === 'number') {
        promise = parser_1.readElf(reader.syncfile(input), options);
    }
    else if (isFileHandle(input)) {
        promise = parser_1.readElf(reader.asyncfile(input), options);
    }
    else if (isBlob(input)) {
        promise = parser_1.readElf(reader.blob(input), options);
    }
    else if (typeof input === 'string') {
        promise = parser_1.readElf(reader.file(input), options);
    }
    else {
        promise = new Promise((resolve) => {
            resolve({
                success: false,
                errors: ['unsupported input type'],
                warnings: []
            });
        });
    }
    if (callback) {
        promise.then(callback);
    }
    return promise;
}
exports.open = open;
__exportStar(require("./reader"), exports);
__exportStar(require("./elf"), exports);
__exportStar(require("./types"), exports);
__exportStar(require("./debug"), exports);
