// elfinfo/dist/src/reader.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.blob = exports.file = exports.syncfile = exports.asyncfile = exports.buffer = exports.array = void 0;
const promises_1 = require("fs/promises");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
function bufferRead(state, length, position) {
    let updatepos = false;
    const { array } = state;
    if (!array) {
        return Promise.reject("state.array must be defined");
    }
    if (!length) {
        return Promise.reject('Length must be specified');
    }
    if (!position) {
        position = state.position;
        updatepos = true;
    }
    if (position <= state.size) {
        if (position + length > state.size) {
            length = state.size - position;
        }
        if (length < 0) {
            length = 0;
        }
        if (updatepos) {
            state.position += length;
        }
        return Promise.resolve(new Uint8Array(array, state.offset + position, length));
    }
    else {
        return Promise.reject("read past end of file");
    }
}
async function asyncFileRead(file, length, position) {
    if (!file)
        throw "File not open";
    const result = new Uint8Array(length);
    const { bytesRead } = await file.read(result, 0, length, position);
    if (bytesRead < length) {
        return result.slice(0, bytesRead);
    }
    else {
        return result;
    }
}
function syncFileRead(file, length, position) {
    return new Promise((resolve, reject) => {
        const result = new Uint8Array(length);
        fs.read(file, result, 0, length, position !== null && position !== void 0 ? position : null, (err, bytesRead) => {
            if (err) {
                reject(err);
            }
            else {
                if (bytesRead < length) {
                    resolve(result.slice(0, bytesRead));
                }
                else {
                    resolve(result);
                }
            }
        });
    });
}
function createView(from) {
    return new DataView(from.buffer, from.byteOffset, from.byteLength);
}
function error_reader(message) {
    return {
        open: () => Promise.reject(message),
        read: (a, b) => Promise.reject(message),
        view: (a, b) => Promise.reject(message),
        size: () => -1,
        close: () => Promise.reject(message)
    };
}
function array(array) {
    return buffer(Uint8Array.from(array));
}
exports.array = array;
function buffer(buffer) {
    const state = {
        array: buffer instanceof Uint8Array ? buffer.buffer : buffer,
        offset: 0,
        position: 0,
        size: buffer.byteLength
    };
    if (buffer instanceof Uint8Array) {
        state.offset = buffer.byteOffset;
    }
    return {
        open: () => Promise.resolve(),
        size: () => state.size,
        close: () => Promise.resolve(),
        read: (length, position) => bufferRead(state, length, position),
        view: (length, position) => bufferRead(state, length, position).then(createView),
    };
}
exports.buffer = buffer;
function asyncfile(fh, ownshandle) {
    if (!fs)
        return error_reader('No filesystem');
    const state = { fh, size: 0, ownshandle };
    return {
        // open checks if the file handle is still valid
        // and gets the size
        open: () => state.fh.stat().then(ss => { state.size = ss.size; }),
        read: (length, position) => asyncFileRead(state.fh, length, position),
        view: (length, position) => asyncFileRead(state.fh, length, position).then(createView),
        size: () => state.size,
        close: () => state.ownshandle ? state.fh.close() : Promise.resolve()
    };
}
exports.asyncfile = asyncfile;
function syncfile(handle, ownshandle) {
    if (!fs)
        return error_reader('No filesystem');
    const state = { ownshandle, handle, size: 0 };
    return {
        open: () => new Promise((resolve, reject) => fs.fstat(state.handle, (e, ss) => {
            // open just checks if the file handle is still valid
            if (e) {
                reject(e);
            }
            else {
                state.size = ss.size;
                resolve();
            }
        })),
        read: (length, position) => syncFileRead(state.handle, length, position),
        view: (length, position) => syncFileRead(state.handle, length, position).then(createView),
        size: () => state.size,
        close: () => new Promise((resolve, reject) => {
            if (state.ownshandle) {
                fs.close(state.handle, e => {
                    if (e) {
                        reject(e);
                    }
                    else {
                        resolve();
                    }
                });
            }
            else {
                resolve();
            }
        })
    };
}
exports.syncfile = syncfile;
function file(elfpath) {
    if (!fs)
        return error_reader('No filesystem');
    const state = { path: elfpath, fh: undefined, size: 0 };
    return {
        open: () => promises_1.open(state.path, 'r')
            .then(fh => {
            state.fh = fh;
            return fh.stat().then(ss => { state.size = ss.size; });
        }),
        read: (length, position) => asyncFileRead(state.fh, length, position),
        view: (length, position) => asyncFileRead(state.fh, length, position).then(createView),
        size: () => state.size,
        close: () => state.fh ? state.fh.close() : Promise.resolve(),
        path: path.resolve(elfpath)
    };
}
exports.file = file;
function blob(item) {
    const state = {
        offset: 0,
        position: 0,
        size: 0
    };
    return {
        open: () => item.arrayBuffer().then(ab => {
            state.array = ab;
            state.offset = 0;
            state.position = 0;
            state.size = ab.byteLength;
        }),
        close: () => Promise.resolve(),
        size: () => state.size,
        read: (length, position) => bufferRead(state, length, position),
        view: (length, position) => bufferRead(state, length, position).then(createView)
    };
}
exports.blob = blob;
