// elfinfo/dist/src/strings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.symbolVisibilityToString = exports.symbolBindingToString = exports.symbolTypeToString = exports.elfFlagsToString = exports.programHeaderFlagsToString = exports.sectionFlagsToString = exports.sectionHeaderEntryTypeToString = exports.programHeaderEntryTypeToString = exports.objectTypeToString = exports.abiToString = exports.isaToString = void 0;
const types_1 = require("./types");
function isaToString(isa) {
    switch (isa) {
        case types_1.ISA.None: return 'No machine';
        case types_1.ISA.M32: return 'AT&T WE 32100';
        case types_1.ISA.SPARC: return 'SUN SPARC';
        case types_1.ISA.x86: return 'Intel x86';
        case types_1.ISA.ISA68K: return 'Motorola m68k family';
        case types_1.ISA.ISA88K: return 'Motorola m88k family';
        case types_1.ISA.IAMCU: return 'Intel MCU';
        case types_1.ISA.ISA860: return 'Intel 80860';
        case types_1.ISA.MIPS: return 'MIPS R3000 big-endian';
        case types_1.ISA.S370: return 'IBM System/370';
        case types_1.ISA.MIPS_RS3_LE: return 'MIPS R3000 little-endian';
        case types_1.ISA.PARISC: return 'HPPA';
        case types_1.ISA.VPP500: return 'Fujitsu VPP500';
        case types_1.ISA.SPARC32PLUS: return 'Sun\'s "v8plus"';
        case types_1.ISA.ISA960: return 'Intel 80960';
        case types_1.ISA.PPC: return 'PowerPC';
        case types_1.ISA.PPC64: return 'PowerPC 64-bit';
        case types_1.ISA.S390: return 'IBM S390';
        case types_1.ISA.SPU: return 'IBM SPU/SPC';
        case types_1.ISA.V800: return 'NEC V800 series';
        case types_1.ISA.FR20: return 'Fujitsu FR20';
        case types_1.ISA.RH32: return 'TRW RH-32';
        case types_1.ISA.RCE: return 'Motorola RCE';
        case types_1.ISA.ARM: return 'ARM';
        case types_1.ISA.FAKE_ALPHA: return 'Digital Alpha';
        case types_1.ISA.SH: return 'Hitachi SH';
        case types_1.ISA.SPARCV9: return 'SPARC v9 64-bit';
        case types_1.ISA.TRICORE: return 'Siemens Tricore';
        case types_1.ISA.ARC: return 'Argonaut RISC Core';
        case types_1.ISA.H8_300: return 'Hitachi H8/300';
        case types_1.ISA.H8_300H: return 'Hitachi H8/300H';
        case types_1.ISA.H8S: return 'Hitachi H8S';
        case types_1.ISA.H8_500: return 'Hitachi H8/500';
        case types_1.ISA.IA_64: return 'Intel Merced';
        case types_1.ISA.MIPS_X: return 'Stanford MIPS-X';
        case types_1.ISA.COLDFIRE: return 'Motorola Coldfire';
        case types_1.ISA.ISA68HC12: return 'Motorola M68HC12';
        case types_1.ISA.MMA: return 'Fujitsu MMA Multimedia Accelerator';
        case types_1.ISA.PCP: return 'Siemens PCP';
        case types_1.ISA.NCPU: return 'Sony nCPU embeeded RISC';
        case types_1.ISA.NDR1: return 'Denso NDR1 microprocessor';
        case types_1.ISA.STARCORE: return 'Motorola Start*Core processor';
        case types_1.ISA.ME16: return 'Toyota ME16 processor';
        case types_1.ISA.ST100: return 'STMicroelectronic ST100 processor';
        case types_1.ISA.TINYJ: return 'Advanced Logic Corp. Tinyj emb.fam';
        case types_1.ISA.X86_64: return 'AMD x86-64 architecture';
        case types_1.ISA.PDSP: return 'Sony DSP Processor';
        case types_1.ISA.PDP10: return 'Digital PDP-10';
        case types_1.ISA.PDP11: return 'Digital PDP-11';
        case types_1.ISA.FX66: return 'Siemens FX66 microcontroller';
        case types_1.ISA.ST9PLUS: return 'STMicroelectronics ST9+ 8/16 mc';
        case types_1.ISA.ST7: return 'STmicroelectronics ST7 8 bit mc';
        case types_1.ISA.ISA68HC16: return 'Motorola MC68HC16 microcontroller';
        case types_1.ISA.ISA68HC11: return 'Motorola MC68HC11 microcontroller';
        case types_1.ISA.ISA68HC08: return 'Motorola MC68HC08 microcontroller';
        case types_1.ISA.ISA68HC05: return 'Motorola MC68HC05 microcontroller';
        case types_1.ISA.SVX: return 'Silicon Graphics SVx';
        case types_1.ISA.ST19: return 'STMicroelectronics ST19 8 bit mc';
        case types_1.ISA.VAX: return 'Digital VAX';
        case types_1.ISA.CRIS: return 'Axis Communications 32-bit emb.proc';
        case types_1.ISA.JAVELIN: return 'Infineon Technologies 32-bit emb.proc';
        case types_1.ISA.FIREPATH: return 'Element 14 64-bit DSP Processor';
        case types_1.ISA.ZSP: return 'LSI Logic 16-bit DSP Processor';
        case types_1.ISA.MMIX: return 'Donald Knuth\'s educational 64-bit proc';
        case types_1.ISA.HUANY: return 'Harvard University machine-independent object files';
        case types_1.ISA.PRISM: return 'SiTera Prism';
        case types_1.ISA.AVR: return 'Atmel AVR 8-bit microcontroller';
        case types_1.ISA.FR30: return 'Fujitsu FR30';
        case types_1.ISA.D10V: return 'Mitsubishi D10V';
        case types_1.ISA.D30V: return 'Mitsubishi D30V';
        case types_1.ISA.V850: return 'NEC v850';
        case types_1.ISA.M32R: return 'Mitsubishi M32R';
        case types_1.ISA.MN10300: return 'Matsushita MN10300';
        case types_1.ISA.MN10200: return 'Matsushita MN10200';
        case types_1.ISA.PJ: return 'picoJava';
        case types_1.ISA.OPENRISC: return 'OpenRISC 32-bit embedded processor';
        case types_1.ISA.ARC_COMPACT: return 'ARC International ARCompact';
        case types_1.ISA.XTENSA: return 'Tensilica Xtensa Architecture';
        case types_1.ISA.VIDEOCORE: return 'Alphamosaic VideoCore';
        case types_1.ISA.TMM_GPP: return 'Thompson Multimedia General Purpose Proc';
        case types_1.ISA.NS32K: return 'National Semi. 32000';
        case types_1.ISA.TPC: return 'Tenor Network TPC';
        case types_1.ISA.SNP1K: return 'Trebia SNP 1000';
        case types_1.ISA.ST200: return 'STMicroelectronics ST200';
        case types_1.ISA.IP2K: return 'Ubicom IP2xxx';
        case types_1.ISA.MAX: return 'MAX processor';
        case types_1.ISA.CR: return 'National Semi. CompactRISC';
        case types_1.ISA.F2MC16: return 'Fujitsu F2MC16';
        case types_1.ISA.MSP430: return 'Texas Instruments msp430';
        case types_1.ISA.BLACKFIN: return 'Analog Devices Blackfin DSP';
        case types_1.ISA.SE_C33: return 'Seiko Epson S1C33 family';
        case types_1.ISA.SEP: return 'Sharp embedded microprocessor';
        case types_1.ISA.ARCA: return 'Arca RISC';
        case types_1.ISA.UNICORE: return 'PKU-Unity & MPRC Peking Uni. mc series';
        case types_1.ISA.EXCESS: return 'eXcess configurable cpu';
        case types_1.ISA.DXP: return 'Icera Semi. Deep Execution Processor';
        case types_1.ISA.ALTERA_NIOS2: return 'Altera Nios II';
        case types_1.ISA.CRX: return 'National Semi. CompactRISC CRX';
        case types_1.ISA.XGATE: return 'Motorola XGATE';
        case types_1.ISA.C166: return 'Infineon C16x/XC16x';
        case types_1.ISA.M16C: return 'Renesas M16C';
        case types_1.ISA.DSPIC30F: return 'Microchip Technology dsPIC30F';
        case types_1.ISA.CE: return 'Freescale Communication Engine RISC';
        case types_1.ISA.M32C: return 'Renesas M32C';
        case types_1.ISA.TSK3000: return 'Altium TSK3000';
        case types_1.ISA.RS08: return 'Freescale RS08';
        case types_1.ISA.SHARC: return 'Analog Devices SHARC family';
        case types_1.ISA.ECOG2: return 'Cyan Technology eCOG2';
        case types_1.ISA.SCORE7: return 'Sunplus S+core7 RISC';
        case types_1.ISA.DSP24: return 'New Japan Radio (NJR) 24-bit DSP';
        case types_1.ISA.VIDEOCORE3: return 'Broadcom VideoCore III';
        case types_1.ISA.LATTICEMICO32: return 'RISC for Lattice FPGA';
        case types_1.ISA.SE_C17: return 'Seiko Epson C17';
        case types_1.ISA.TI_C6000: return 'Texas Instruments TMS320C6000 DSP';
        case types_1.ISA.TI_C2000: return 'Texas Instruments TMS320C2000 DSP';
        case types_1.ISA.TI_C5500: return 'Texas Instruments TMS320C55x DSP';
        case types_1.ISA.TI_ARP32: return 'Texas Instruments App. Specific RISC';
        case types_1.ISA.TI_PRU: return 'Texas Instruments Prog. Realtime Unit';
        case types_1.ISA.MMDSP_PLUS: return 'STMicroelectronics 64bit VLIW DSP';
        case types_1.ISA.CYPRESS_M8C: return 'Cypress M8C';
        case types_1.ISA.R32C: return 'Renesas R32C';
        case types_1.ISA.TRIMEDIA: return 'NXP Semi. TriMedia';
        case types_1.ISA.QDSP6: return 'QUALCOMM DSP6';
        case types_1.ISA.ISA8051: return 'Intel 8051 and variants';
        case types_1.ISA.STXP7X: return 'STMicroelectronics STxP7x';
        case types_1.ISA.NDS32: return 'Andes Tech. compact code emb. RISC';
        case types_1.ISA.ECOG1X: return 'Cyan Technology eCOG1X';
        case types_1.ISA.MAXQ30: return 'Dallas Semi. MAXQ30 mc';
        case types_1.ISA.XIMO16: return 'New Japan Radio (NJR) 16-bit DSP';
        case types_1.ISA.MANIK: return 'M2000 Reconfigurable RISC';
        case types_1.ISA.CRAYNV2: return 'Cray NV2 vector architecture';
        case types_1.ISA.RX: return 'Renesas RX';
        case types_1.ISA.METAG: return 'Imagination Tech. META';
        case types_1.ISA.MCST_ELBRUS: return 'MCST Elbrus';
        case types_1.ISA.ECOG16: return 'Cyan Technology eCOG16';
        case types_1.ISA.CR16: return 'National Semi. CompactRISC CR16';
        case types_1.ISA.ETPU: return 'Freescale Extended Time Processing Unit';
        case types_1.ISA.SLE9X: return 'Infineon Tech. SLE9X';
        case types_1.ISA.L10M: return 'Intel L10M';
        case types_1.ISA.K10M: return 'Intel K10M';
        case types_1.ISA.AARCH64: return 'ARM AARCH64';
        case types_1.ISA.AVR32: return 'Amtel 32-bit microprocessor';
        case types_1.ISA.STM8: return 'STMicroelectronics STM8';
        case types_1.ISA.TILE64: return 'Tileta TILE64';
        case types_1.ISA.TILEPRO: return 'Tilera TILEPro';
        case types_1.ISA.MICROBLAZE: return 'Xilinx MicroBlaze';
        case types_1.ISA.CUDA: return 'NVIDIA CUDA';
        case types_1.ISA.TILEGX: return 'Tilera TILE-Gx';
        case types_1.ISA.CLOUDSHIELD: return 'CloudShield';
        case types_1.ISA.COREA_1ST: return 'KIPO-KAIST Core-A 1st gen.';
        case types_1.ISA.COREA_2ND: return 'KIPO-KAIST Core-A 2nd gen.';
        case types_1.ISA.ARC_COMPACT2: return 'Synopsys ARCompact V2';
        case types_1.ISA.OPEN8: return 'Open8 RISC';
        case types_1.ISA.RL78: return 'Renesas RL78';
        case types_1.ISA.VIDEOCORE5: return 'Broadcom VideoCore V';
        case types_1.ISA.ISA78KOR: return 'Renesas 78KOR';
        case types_1.ISA.ISA56800EX: return 'Freescale 56800EX DSC';
        case types_1.ISA.BA1: return 'Beyond BA1';
        case types_1.ISA.BA2: return 'Beyond BA2';
        case types_1.ISA.XCORE: return 'XMOS xCORE';
        case types_1.ISA.MCHP_PIC: return 'Microchip 8-bit PIC(r)';
        case types_1.ISA.KM32: return 'KM211 KM32';
        case types_1.ISA.KMX32: return 'KM211 KMX32';
        case types_1.ISA.EMX16: return 'KM211 KMX16';
        case types_1.ISA.EMX8: return 'KM211 KMX8';
        case types_1.ISA.KVARC: return 'KM211 KVARC';
        case types_1.ISA.CDP: return 'Paneve CDP';
        case types_1.ISA.COGE: return 'Cognitive Smart Memory Processor';
        case types_1.ISA.COOL: return 'Bluechip CoolEngine';
        case types_1.ISA.NORC: return 'Nanoradio Optimized RISC';
        case types_1.ISA.CSR_KALIMBA: return 'CSR Kalimba';
        case types_1.ISA.Z80: return 'Zilog Z80';
        case types_1.ISA.VISIUM: return 'Controls and Data Services VISIUMcore';
        case types_1.ISA.FT32: return 'FTDI Chip FT32';
        case types_1.ISA.MOXIE: return 'Moxie processor';
        case types_1.ISA.AMDGPU: return 'AMD GPU';
        case types_1.ISA.RISCV: return 'RISC-V';
        case types_1.ISA.BPF: return 'Linux BPF';
        case types_1.ISA.CSKY: return 'C-SKY';
    }
}
exports.isaToString = isaToString;
function abiToString(abi) {
    switch (abi) {
        case types_1.ABI.GNUHurd: return 'GNU Hurd';
        case types_1.ABI.NovelloModesto: return 'Novello Modesto';
        case types_1.ABI.HPUX: return 'HP-UX';
        case types_1.ABI.NonStopKernel: return 'NonStop Kernel';
        case types_1.ABI.FenixOS: return 'Fenix OS';
        case types_1.ABI.ARMEABI: return 'ARM EABI';
        default:
            return types_1.ABI[abi] || abi.toString();
    }
}
exports.abiToString = abiToString;
function objectTypeToString(objectType) {
    return types_1.ObjectType[objectType] || objectType.toString();
}
exports.objectTypeToString = objectTypeToString;
function programHeaderEntryTypeToString(programHeaderEntryType) {
    switch (programHeaderEntryType) {
        case types_1.ProgramHeaderEntryType.ProgramHeaderTable:
            return 'Program Header Table';
        case types_1.ProgramHeaderEntryType.GnuEhFrame:
            return 'GNU EH frame';
        case types_1.ProgramHeaderEntryType.GnuStack:
            return 'GNU stack info';
        case types_1.ProgramHeaderEntryType.GnuRelRo:
            return 'GNU ro relocation';
        default:
            return types_1.ProgramHeaderEntryType[programHeaderEntryType] || programHeaderEntryType.toString();
    }
}
exports.programHeaderEntryTypeToString = programHeaderEntryTypeToString;
function sectionHeaderEntryTypeToString(sectionHeaderEntryType) {
    switch (sectionHeaderEntryType) {
        case types_1.SectionHeaderEntryType.Null: return 'NULL';
        case types_1.SectionHeaderEntryType.ProgBits: return 'Prog bits';
        case types_1.SectionHeaderEntryType.SymTab: return 'Symbol table';
        case types_1.SectionHeaderEntryType.StrTab: return 'String table';
        case types_1.SectionHeaderEntryType.Rela: return 'Relocation with addends';
        case types_1.SectionHeaderEntryType.Hash: return 'Symbol hash table';
        case types_1.SectionHeaderEntryType.Dynamic: return 'Dynamic';
        case types_1.SectionHeaderEntryType.Note: return 'Note';
        case types_1.SectionHeaderEntryType.NoBits: return 'No bits';
        case types_1.SectionHeaderEntryType.Rel: return 'Relocation';
        case types_1.SectionHeaderEntryType.ShLib: return 'ShLib';
        case types_1.SectionHeaderEntryType.DynSym: return 'Dynamic linking symbols';
        case types_1.SectionHeaderEntryType.InitArray: return 'Init array';
        case types_1.SectionHeaderEntryType.FiniArray: return 'Fini array';
        case types_1.SectionHeaderEntryType.PreInitArray: return 'Pre-init array';
        case types_1.SectionHeaderEntryType.Group: return 'Section group';
        case types_1.SectionHeaderEntryType.ShNdx: return 'Extended symbol table index';
        case types_1.SectionHeaderEntryType.Num: return 'Num';
        case types_1.SectionHeaderEntryType.GnuAttributes: return 'GNU object attributes';
        case types_1.SectionHeaderEntryType.GnuHash: return 'GNU hash';
        case types_1.SectionHeaderEntryType.GnuLibList: return 'GNU pre-link library list';
        case types_1.SectionHeaderEntryType.GnuVerDef: return 'GNU version definition';
        case types_1.SectionHeaderEntryType.GnuVerNeed: return 'GNU version needs';
        case types_1.SectionHeaderEntryType.GnuVerSym: return 'GNU version symbol table';
        default:
            return types_1.SectionHeaderEntryType[sectionHeaderEntryType] || Number(sectionHeaderEntryType).toString();
    }
}
exports.sectionHeaderEntryTypeToString = sectionHeaderEntryTypeToString;
function sectionFlagsToString(flags) {
    // no flags are more than 32 bits
    flags = Number(flags);
    let str = [];
    if (flags & 0x1)
        str.push('Writeable');
    if (flags & 0x2)
        str.push('Alloc');
    if (flags & 0x4)
        str.push('Executable');
    if (flags & 0x10)
        str.push('Merge');
    if (flags & 0x20)
        str.push('Strings');
    if (flags & 0x40)
        str.push('Info Link');
    if (flags & 0x80)
        str.push('Link Order');
    if (flags & 0x100)
        str.push('Nonconforming');
    if (flags & 0x200)
        str.push('Group');
    if (flags & 0x400)
        str.push('Thread Local Storage');
    if (flags & 0x4000000)
        str.push('Special ordering');
    if (flags & 0x8000000)
        str.push('Exclude');
    if (str.length == 0)
        return '<none>';
    return str.join(' | ');
}
exports.sectionFlagsToString = sectionFlagsToString;
function programHeaderFlagsToString(flags) {
    let str = [];
    if (flags & 0x4)
        str.push('Read');
    if (flags & 0x2)
        str.push('Write');
    if (flags & 0x1)
        str.push('Execute');
    return str.join(' | ');
}
exports.programHeaderFlagsToString = programHeaderFlagsToString;
function elfFlagsToString(isa, flags) {
    if (isa === types_1.ISA.ARM) {
        const ver = ((flags & 0xff000000) >> 24);
        let str = [
            `Version: ${ver}`
        ];
        if (flags & 0x00800000)
            str.push('BE-8');
        if (ver <= 4 && flags & 0x00400000)
            str.push('Legacy');
        if (ver >= 5 && flags & 0x00000400)
            str.push('Hard Float');
        if (ver >= 5 && flags & 0x00000200)
            str.push('Soft Float');
        return str.join(' | ');
    }
    return flags.toString();
}
exports.elfFlagsToString = elfFlagsToString;
function symbolTypeToString(type) {
    switch (type) {
        case types_1.SymbolType.RelocationExpression: return "Relocation Expression";
        case types_1.SymbolType.SignedRelocationExpression: return "Signed Relocation Expression";
        case types_1.SymbolType.ThreadLocalStorage: return "Thread Local Storage";
        default: return types_1.SymbolType[type] || type.toString();
    }
}
exports.symbolTypeToString = symbolTypeToString;
function symbolBindingToString(binding) {
    return types_1.SymbolBinding[binding] || binding.toString();
}
exports.symbolBindingToString = symbolBindingToString;
function symbolVisibilityToString(visibility) {
    return types_1.SymbolVisibility[visibility] || visibility.toString();
}
exports.symbolVisibilityToString = symbolVisibilityToString;
