// elfinfo/dist/tests/index.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.test = exports.category = void 0;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const async_hooks_1 = require("async_hooks");
const perf_hooks_1 = require("perf_hooks");
const { stdout, stderr, argv } = process;
let passed = 0, failed = 0;
let totaltime = 0;
const startTime = perf_hooks_1.performance.now();
const tests = [];
const currentCategoryStorage = new async_hooks_1.AsyncLocalStorage();
const categoryPromises = [];
function getErrorMessage(e) {
    var _a;
    return e ? ((_a = e.stack) !== null && _a !== void 0 ? _a : (e.toString ? e.toString() : `${e}`)) : undefined;
}
function printErrorMessage(message) {
    if (message) {
        message = message.split('\n').map(l => `    ${l}`).join('\n');
        stderr.write(message, 'utf-8');
        stderr.write('\n', 'utf-8');
    }
}
function getTimeString(time) {
    return time !== undefined && time > 100 ? ` (${time.toFixed(0)}ms)` : '';
}
function printStarting(name) {
    stdout.write(`👉 ${name}`, 'utf-8');
}
function printSuccess(name, time) {
    stdout.write(`\r✔️  ${name}${getTimeString(time)}\n`, 'utf-8');
}
function printFailure(name, message, time) {
    stderr.write(`\r❌ ${name}${getTimeString(time)}\n`, 'utf-8');
    printErrorMessage(message);
}
function category(name, stuff) {
    const prev = currentCategoryStorage.getStore();
    if (prev) {
        name = `${prev.name} - ${name}`;
    }
    const promise = currentCategoryStorage.run({ name, prev }, stuff);
    if (promise) {
        categoryPromises.push({ promise, name });
    }
}
exports.category = category;
function test(name, fn) {
    const category = currentCategoryStorage.getStore();
    if (category) {
        name = `${category.name} - ${name}`;
    }
    tests.push({ name, fn });
}
exports.test = test;
async function findTests() {
    const files = (await promises_1.default.readdir(__dirname))
        .filter(x => /^(?!index).+\.js$/.test(x));
    for (let test of files) {
        require(path_1.default.join(__dirname, test));
    }
    await Promise.allSettled(categoryPromises.map(({ promise, name }) => promise.catch(e => {
        printFailure(name, getErrorMessage(e));
        failed++;
    })));
}
async function run() {
    const filter = argv[2] ? new RegExp(argv[2]) : undefined;
    const timeUntilRun = perf_hooks_1.performance.now() - startTime;
    totaltime += timeUntilRun;
    const timeUntilRunStr = getTimeString(timeUntilRun);
    if (timeUntilRunStr) {
        stdout.write(`Bootstrapped tests${timeUntilRunStr}\n`, 'utf-8');
    }
    for (const { name, fn } of tests) {
        if (!filter || filter.test(name)) {
            let success = false;
            let errorMessage;
            let start = 0;
            let time = 0;
            try {
                printStarting(name);
                stdout.uncork();
                start = perf_hooks_1.performance.now();
                const result = fn();
                if (result instanceof Promise) {
                    await result;
                }
                time = perf_hooks_1.performance.now() - start;
                success = true;
            }
            catch (e) {
                time = perf_hooks_1.performance.now() - start;
                success = false;
                errorMessage = getErrorMessage(e);
            }
            stdout.cork();
            stderr.cork();
            if (success) {
                printSuccess(name, time);
                passed++;
            }
            else {
                printFailure(name, errorMessage, time);
                failed++;
            }
            stdout.uncork();
            stderr.uncork();
        }
    }
    if (failed) {
        stderr.write(`\n \x1b[1;5;91mFAILED\x1b[0m (\x1b[1;97m${failed}\x1b[0m tests failed, \x1b[1;97m${passed}\x1b[0m tests passed in ${(totaltime / 1000).toFixed(2)}s)\n`);
        process.exit(1);
    }
    else {
        stdout.write(`\n \x1b[1;5;92mPASSED\x1b[0m (\x1b[1;97m${passed}\x1b[0m tests passed in ${(totaltime / 1000).toFixed(2)}s)\n`);
        process.exit(0);
    }
}
(async () => {
    await findTests();
    await run();
})();
