// elfinfo/dist/tests/testprograms.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.specialPrograms = exports.stockPrograms = exports.testPrograms = exports.stockProgramAbis = exports.stockProgramNames = exports.testProgramAbis = exports.testProgramObjects = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const basepath = path.resolve(path.join(__dirname, '..', '..', 'testprograms'));
exports.testProgramObjects = ['factorial', 'main', 'startup', 'syscalls', 'vectortable'];
exports.testProgramAbis = ['arm-eabi', 'clang-x64', 'gcc-x64', 'riscv-eabi'];
exports.stockProgramNames = ['cat', 'date'];
exports.stockProgramAbis = ['alpha', 'amd64', 'armhf', 'hppa',
    'hurd', 'i386', 'ia64', 'm68k', 'powerpc', 'ppc64', 'riscv64', 'sh4', 'sparc64', 'x32'];
function progpath(abi) {
    return path.resolve(path.join(basepath, 'bin', abi, 'program'));
}
function objpaths(abi) {
    return exports.testProgramObjects.map(object => path.resolve(path.join(basepath, 'obj', abi, `${object}.o`)));
}
function read(path) {
    return fs.readFileSync(path);
}
function getTestPrograms() {
    return Object.fromEntries(exports.testProgramAbis.map(abi => ({
        abi,
        vals: {
            objectPaths: objpaths(abi),
            programPath: progpath(abi)
        }
    })).map(({ abi, vals: { objectPaths, programPath } }) => [abi, {
            objectPaths: objectPaths,
            objects: objectPaths.map(read),
            programPath: programPath,
            program: read(programPath),
        }]));
}
function getStockPrograms() {
    return Object.fromEntries(exports.stockProgramNames.map(name => [name, Object.fromEntries(exports.stockProgramAbis.map(abi => {
            const programPath = path.join(basepath, 'stock', `${name}_${abi}`);
            return [abi, {
                    name,
                    abi,
                    program: read(programPath),
                    programPath
                }];
        }))]));
}
function getSpecialPrograms() {
    const dir = path.join(basepath, 'special');
    return fs.readdirSync(dir)
        .map(n => ({ name: n, path: path.join(dir, n) }));
}
exports.testPrograms = getTestPrograms();
exports.stockPrograms = getStockPrograms();
exports.specialPrograms = getSpecialPrograms();
